libs/beacon/src/beacon_manager.cpp

```diff
@@ -17,167 +17,617 @@
+      {
+        bn::G2 rhs, lhs;
+        lhs = crypto::mcl::ComputeLHS(g__s_ij[i][cabinet_index_], group_g_, group_h_,
+                                      s_ij[i][cabinet_index_], sprime_ij[i][cabinet_index_]);
+        rhs = crypto::mcl::ComputeRHS(cabinet_index_, C_ik[i]);
+        if (lhs != rhs)
+        {
+          FETCH_LOG_WARN(LOGGING_NAME, "Node ", cabinet_index_,
+                         " received bad coefficients/shares from node ", i);
+          complaints_local.insert(cab.first);
+        }
+      }
+      else
+      {
+        FETCH_LOG_WARN(LOGGING_NAME, "Node ", cabinet_index_,
+                       " received vanishing coefficients/shares from ", i);
+        complaints_local.insert(cab.first);
+      }
+    }
+    ++i;
   }
+  return complaints_local;
+}
 
-  uint64_t n = it->second;
-  return contribution_.contributions[n];
+bool BeaconManager::VerifyComplaintAnswer(MuddleAddress const &from, ComplaintAnswer const &answer)
+{
+  CabinetIndex from_index = identity_to_index_[from];
+  if (identity_to_index_.find(answer.first) == identity_to_index_.end())
+  {
+    FETCH_LOG_WARN(LOGGING_NAME, "Node ", cabinet_index_,
+                   " received complaint answer with unknown reporter index");
+    return true;
+  }
+
+  CabinetIndex reporter_index = identity_to_index_[answer.first];
+  // Verify shares received
+  bn::Fr s, sprime;
+  bn::G2 lhsG, rhsG;
+  s.clear();
+  sprime.clear();
+  lhsG.clear();
+  rhsG.clear();
+  s.setStr(answer.second.first);
+  sprime.setStr(answer.second.second);
+  rhsG = crypto::mcl::ComputeRHS(reporter_index, C_ik[from_index]);
+  lhsG = crypto::mcl::ComputeLHS(group_g_, group_h_, s, sprime);
+  if (lhsG != rhsG)
+  {
+    FETCH_LOG_WARN(LOGGING_NAME, "Node: ", cabinet_index_, " verification for node ", from_index,
+                   " complaint answer failed");
+    return false;
+  }
+  else
+  {
+    FETCH_LOG_INFO(LOGGING_NAME, "Node: ", cabinet_index_, " verification for node ", from_index,
+                   " complaint answer succeeded");
+    if (reporter_index == cabinet_index_)
+    {
+      FETCH_LOG_INFO(LOGGING_NAME, "Node: ", cabinet_index_, " reset shares for ", from_index);
+      s_ij[from_index][cabinet_index_]      = s;
+      sprime_ij[from_index][cabinet_index_] = sprime;
+      g__s_ij[from_index][cabinet_index_].clear();
+      bn::G2::mul(g__s_ij[from_index][cabinet_index_], group_g_, s_ij[from_index][cabinet_index_]);
+    }
+    return true;
+  }
 }
 
-bool BeaconManager::AddShare(BeaconManager::Identity from, BeaconManager::PrivateKey share,
-                             BeaconManager::VerificationVector verification)
+/**
+ * If in qual a member computes individual share of the secret key and further computes and
+ * broadcasts qual coefficients
+ */
+void BeaconManager::ComputeSecretShare()
 {
-  auto it = identity_to_index_.find(from);
-  assert(it != identity_to_index_.end());
+  secret_share_.clear();
+  xprime_i = 0;
```

Comment:
is = 0 here the same as clear? It doesn't need to do it on every iteration of the qual loop?